### ISSUE TITLE
Why?

### ISSUE THREAD

**Issue by orefalo** [mologie/meteor-typescript#1]
Good morning,

My name is Olivier, I am one of the contributors at https://atmospherejs.com/meteortypescript

With all due respect, I don't understand the benefits of your module. Don't take me wrong, you did a good job at integrating with the Typescript compiler API. But from a functional standpoint, your module is a subset of https://atmospherejs.com/meteortypescript/compiler.

Can you explain the drivers behind your contribution?

**Comment by mologie** [comment]
Hello Olivier,

thank you for your honest question - this will be an interesting discussion.

I originally attempted to use your module in an application of mine, and actually did mention your module in this project's readme file. I ended up writing this module instead of using yours because I believe that your approach for integrating TypeScript into Meteor is not future-proof, and due to the large amount of hacks and workarounds not ready for a production application.

Meteor's build tool expects at least one output file per input file. It does not officially support batch-processing input files, and probably never will. Batch processing as it is available in your module is in my opinion an unnecessary complication. It results in a small speed advantage, but requires an awkward workaround. The Meteor build tool may support incremental application builds soon (the MDG is discussing this on their Hackpad account), so that batch processing becomes unnecessary.

I was not able to properly get your module to work with Meteor packages either. Due to the workaround placeholder file that is required by your module, it is not possible to just run a clean Meteor project which makes use of packages that depend on your module. Meteor has to be started multiple times for all placeholder files to be created. This breaks automated unit testing and was thus a major roadblock for me.

Meteor is also quickly approaching version 1.1, which will come with Windows support. Windows has its own quirks, but it is a commonly used server platform that I want to target with my application. I was therefore looking for a solution that would work on Windows out of the box.

I hope that I was able to explain my motivation for writing this module to you. I would have prefered to support the development of your module, but the concepts and approach that I would have taken were entirely different from how your module already worked, and I felt that it would be more appropriate to begin a new project which fits my ideas instead.

Oliver

**Comment by orefalo** [comment]
Good evening Oliver,

I see plenty of good points in your reply. Let me add my feedback to the mix. 

First let me state that the previous version of typescript-compiler was working exactly like yours. At the time, there was no standard Typescript compiler API, so I had to compose with 3rd party npm modules..

Here are the issues I experienced
- Slow compilation - there is no doubt that a batch compilation is faster that file by file. You can test it from the command line with a real TS project. Unlike Coffeescript and other transpilers, Typescript typically loads a bunch of reference file s(definitions) to do static typing. We are talking a factor 10x here.
- Clean separation of definition and implementation - directly related to the above, compiling a project 'file by file' requires a clean separation of concerns: definition file (.d.ts) vs implementation (.ts). Doing otherwise generates duplicate code blocks. This is not a bad practice by itself, but JS coders are not used to it - Indeed it makes TS programming much heavier.
- Memory leaks  - Last but not least, instantiating a compiler is the same memory space as the application is definitely not good practice.

To your points:
- Windows is not supported - meteor windows is still in its preliminary phases. as long as the compiler can be called with the tsc command line, it should work.
- Meteor packages - please provide more details by opening an issue on the project. it will be fixed.
- Future proof - Meteor is continuously evolving, there is no such a thing as future proof in this ecosystem. Should meteor change the way it compile files, your implementation will need a rewrite as well.  It happened with the release of 1.0 and will happen again. My point is future proof = support.

Bottom line, it looks like both implementations have benefits and shortcomings, maybe the meteor prj needs to do something about it...
But I find that there is no need to split an already small community around two TS compilers (back to my point about support). If you want a different compiler behavior, do your implementation on top of the existing one. It only takes a flag/property file, to have different implementations. Besides I always give full privileges to the contributors who request it - so you won't "loose" anything.

Sincerely,

**Comment by mologie** [comment]
Hello Olivier,

please excuse my delayed response - last week was a busy week. I have also had to deal with some of the issue that you presented. Please let me explain my thoughts and solutions:

Profiling has shown that slow compilation is mainly caused by the constant reparsing of all referenced files (mainly lib.d.ts, which weights half a megabyte). Microsoft experienced the same issue when integrating TypeScript into Visual Studio and for this reason separated the syntax tree generation from the actual type checking and file emission steps.

I am now leveraging the TypeScript compiler's structure in order to provide various caching machanisms that significantly speed up compilation:
- A syntax tree (AST) cache skips redundant reparsing of input files (including lib.d.ts). When initially processing a TypeScript project file-by-file, the AST cache is filled with files as they are parsed. I store the file's timestamps in order to transparently invalidate the cache if any referenced file change.
- In addition to the AST cache, I have introduced a JavaScript/Map cache in order to combat Meteor's lack of incremental file processing. When Meteor asks my plugin to recompile a single file, and neither that file nor its dependencies have been modified, my plugin will return the previous compilation result at the cost of only a few stat syscalls.
- By combining AST cache and JS cache, I was able to achieve compilation speeds nearly equal to batch compilation on the first compiler run (when Meteor starts), and significantly faster than batch compilation when single files change. Changes to HTML and CSS files are now transmitted to the client with near zero delay.

Duplicate code blocks are only generated when compiling a project file-by-file while TypeScript's bundler is active. The bundler is for example activated when using the `--out` command line switch of `tsc`. TypeScript's compiler API by default does not use the bundler and does therefore not generate duplicate code blocks when referencing `.ts` files.

There is slight overhead connected to the method that I currently use: TypeScript generates JavaScript files for all files which the target file references. My plugin filters these duplicates for now and only forwards the actual target file to Meteor. This API issue has already been fixed in the upcoming TypeScript version 1.5.

A valid disadvantage of compiling file-by-file is that TypeScript files of the same package/application do not implicitly reference eachother. Visual Studio used to behave like this some time ago. Microsoft addressed the issue by adding implicit reference support for incremental builds to the TypeScript Language Services API. However, it is not possible to make use of this feature until Meteor provides build plugins with a full list of files to be compiled. I would consider this issue to be more of a slight inconvenience though - it does not have any consequences other than having to add a reference comment to each source file.

In repsonse to your other points:
- Memory leaks - Agreed, good practice it is not. TypeScript's developers also agree and are for this reason adding a `tsserver` tool to TypeScript 1.5, which allows separating the compilation and cache process from Visual Studio (and probably also Meteor.) I would not consider memory leaks too much of an issue though: The Meteor development environment is restarted often, and build plugins do not run in production environments.
- Windows support is actually right around the corner: https://github.com/meteor/meteor/wiki/Preview-of-Meteor-on-Windows This is indeed not a major issue in neither of our plugins.
- Packages / issues with unit testing - This is a conceptual issue. Your plugin needs to ensure that a placeholder file exists, but cannot add this placeholder file to Meteor's list of files to be built right away. It therefore issues an error message as workaround, which breaks automated builds. The fix was trivial, but is up to the user: Check the placeholder file into the source control system.
- Future proof - You are absolutely right, and this is something that I should not have mentioned.

Regarding cooperation and/or merging our plugins:

I do not believe that competition, if one may call it that way with open-source projects, is a bad thing. I may never have investigated the caching mechanisms that my plugin now employs if it were not for our discussion here. The community profits from both our work.

Merging our plugins is as of now contraproductive: Should the user want to select a different compilation method, then adding a different plugin is the most simple solution for the user. Our plugins are after all not compatible: Yours provides implicit referencing and explicit declaration of exports, mine requires explicit referencing but implcitily exports classes and modules. A project may house both our plugins as long as they are not used by the same package.

Once Meteor provides a new build plugin API, I will gladly reconsider colaborating on a TypeScript plugin for Meteor. Implicit references, like it is done in your plugin and in the latest Visual Studio version, is definitely the better solution.

Oliver

**Comment by pgrm** [comment]
Hi, maybe this is not the right topic, just wanted to write it here as both of you are going to read it:

I started off initially with the typescript-compiler https://atmospherejs.com/meteortypescript/compiler - however, it doesn't really work with internal modules. So I had to use `declare var ...` to share data and logic between my files. This doesn't work with expanding classes though.

After I found this compiler, I was skeptical at first, as  it is very new and not used very much, but I gave it a try and now TypeScript works as it's supposed to - it's awesome, and btw - the bug I've reported with files called `main.ts` and `main.anything.ts`: https://github.com/meteor-typescript/meteor-typescript-compiler/issues/23 doesn't show up here. 

just my 2 cents,
Peter

**Comment by orefalo** [comment]
I think you are right pgrm,
My apologies; I don't have the time to fix the issues on the ts compiler at this time. My idea was to merge the two compilers under the same umbrella. I failed to get momentum from the author.
